middleware/rollup.go

```diff
@@ -0,0 +1,164 @@
+package middleware
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"gopkg.in/Clever/kayvee-go.v6/logger"
+)
+
+var globalRollupRouter *RollupRouter
+
+// RollupLogger will log info / error rollups depending on status code.
+type RollupLogger interface {
+	InfoD(title string, data map[string]interface{})
+	ErrorD(title string, data map[string]interface{})
+}
+
+// EnableRollups turns on rollups for kv middleware logs.
+func EnableRollups(ctx context.Context, logger RollupLogger) {
+	globalRollupRouter = NewRollupRouter(ctx, logger)
+}
+
+// RollupRouter rolls up log lines and periodically logs them as one log line.
+type RollupRouter struct {
+	logger  RollupLogger
+	ctx     context.Context
+	ctxDone bool
+
+	// create a rollup object per unique (status-code, path) pair
+	rollupsMu sync.Mutex
+	rollups   map[string]*logRollup
+}
+
+// NewRollupRouter creates a new log rollup output.
+// Rollups will stop when the context is canceled.
+func NewRollupRouter(ctx context.Context, logger RollupLogger) *RollupRouter {
+	l := &RollupRouter{
+		logger:  logger,
+		rollups: map[string]*logRollup{},
+		ctx:     ctx,
+		ctxDone: false,
+	}
+	go func() {
+		select {
+		case <-ctx.Done():
+			l.rollupsMu.Lock()
+			l.rollups = map[string]*logRollup{}
+			l.ctxDone = true
+			l.rollupsMu.Unlock()
+		}
+	}()
+	return l
+}
+
+// Process a log message to roll up.
+func (r *RollupRouter) Process(logmsg map[string]interface{}) {
+	if r.ctxDone {
+		return
+	}
+
+	statusCode, ok := logmsg["status-code"].(int)
+	if !ok {
+		return
+	}
+	path, ok := logmsg["path"].(string)
+	if !ok {
+		return
+	}
+	canary, ok := logmsg["canary"].(bool)
+	if !ok {
+		return
+	}
+	r.findOrCreate(statusCode, path, canary).add(logmsg)
+}
+
+func (r *RollupRouter) findOrCreate(statusCode int, path string, canary bool) *logRollup {
+	r.rollupsMu.Lock()
+	defer r.rollupsMu.Unlock()
+	rollupKey := fmt.Sprintf("%d-%s", statusCode, path)
+	if canary {
+		rollupKey += "-canary"
+	}
+	if rollup, ok := r.rollups[rollupKey]; ok {
+		return rollup
+	}
+	rollup := &logRollup{
+		Logger:           r.logger,
+		StatusCode:       statusCode,
+		Path:             path,
+		Canary:           canary,
+		ReportingDelayNs: (time.Second * 20).Nanoseconds(), // todo: make configurable
+	}
+	r.rollups[rollupKey] = rollup
+	go rollup.schedule(r.ctx)
+	return rollup
+}
+
+// logRollup represents a single rollup.
+type logRollup struct {
+	Logger           RollupLogger
+	StatusCode       int
+	Path             string
+	Canary           bool
+	ReportingDelayNs int64
+
+	rollupMsgMu sync.Mutex
+	rollupMsg   map[string]interface{}
+}
+
+func (r *logRollup) report() {
+	r.rollupMsgMu.Lock()
+	defer r.rollupMsgMu.Unlock()
+	if r.rollupMsg != nil {
+		switch logLevelFromStatus(r.StatusCode) {
+		case logger.Error:
+			r.Logger.ErrorD("request-finished", r.rollupMsg)
+		default:
+			r.Logger.InfoD("request-finished", r.rollupMsg)
```

Comment:
I wonder if we should change the title, e.g. `request-finished-rollup`? that way people don't mistakenly try to join this data with other request finished data. 

Alternately, normal `request-finished` logs could have the same fields, just count = 1

Reply:
509e68e